### PR TITLE
client: fix cancel watch

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/2669

### CODE REVIEW

**Comment by xiang90** [comment]
Manually tested.

File: client/client_test.go [hunk 1]
```diff
@@ -218,6 +221,48 @@ func TestSimpleHTTPClientDoCancelContextResponseBodyClosed(t *testing.T) {
...
+	c chan struct{}
+}
+
+func (bb *blockingBody) Read(p []byte) (n int, err error) {
+	<-bb.c
+	return 0, errors.New("closed")
+}
+
+func (bb *blockingBody) Close() error {
+	bb.c <- struct{}{}
```
Comment (yichengq) [inline client/client_test.go]: nit: close(bb.c)

File: client/client_test.go [hunk 2]
```diff
@@ -218,6 +221,48 @@ func TestSimpleHTTPClientDoCancelContextResponseBodyClosed(t *testing.T) {
...
+
+	// create an already-cancelled context
+	ctx, cancel := context.WithCancel(context.Background())
+	body := &checkableReadCloser{ReadCloser: &blockingBody{c: make(chan struct{})}}
+	go func() {
+		tr.respchan <- &http.Response{Body: body}
+		time.Sleep(2 * time.Millisecond)
+		// cancel after the body is received
+		cancel()
+		// wait for CancelRequest to be called, informing us that simpleHTTPClient
```
Comment (yichengq) [inline client/client_test.go]: we don't expect to call CancelRequest, right? We only expect that the body to be closed.

**Comment by xiang90** [comment]
@yichengq PTAL

File: client/client_test.go [hunk 3]
```diff
@@ -218,6 +221,44 @@ func TestSimpleHTTPClientDoCancelContextResponseBodyClosed(t *testing.T) {
...
+func (bb *blockingBody) Close() error {
+	close(bb.c)
+	return nil
+}
+
+func TestSimpleHTTPClientDoCancelContextResponseBodyClosedWithBlockingBody(t *testing.T) {
+	tr := newFakeTransport()
+	c := &simpleHTTPClient{transport: tr}
+
+	// create an already-cancelled context
```
Comment (yichengq) [inline client/client_test.go]: hrm. not an already-cancelled context here.

**Comment by yichengq** [comment]
lgtm. defer to @bcwaldon if he has interest.

File: client/client.go [hunk 4]
```diff
@@ -293,7 +293,23 @@ func (c *simpleHTTPClient) Do(ctx context.Context, act httpAction) (*http.Respon
...
+	go func() {
+		body, err = ioutil.ReadAll(resp.Body)
+		done <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		err = resp.Body.Close()
+		<-done
+		if err != nil {
```
Comment (bcwaldon) [inline client/client.go]: should this be `if err == nil`?

File: client/client_test.go (hunk 3 above)
Comment (bcwaldon) [inline client/client_test.go]: yeah, don't you need to call `cancel()` before you use `ctx` below?

Comment (xiang90) [inline client/client_test.go]: should not copy code next time. :(

**Pull request merged**